[CPU] Move to oneDNN v3.1

EgorDuplensky opened the pull request:
### Details:
- Operational descs have been removed from oneDNN, so the code have been updated accordingly
- Refactored primitive creation procedure for nodes which use oneDNN primitives
- Sparse desc interface refactored with respect to oneDNN updates
- Minor refactoring

### Migration branch:
- https://github.com/openvinotoolkit/oneDNN/tree/v3.1_for_ie_master_migration

### Tickets:
- 89807

### TBD:
- some sse41 Softmax tests fail on Windows

EgorDuplensky commented:
@dmitry-gorokhov Could you please take a look?
Few issues left, no major changes are expected

EgorDuplensky commented on src/plugins/intel_cpu/src/node.cpp:
```diff
@@ -771,10 +737,67 @@ void Node::initDescriptor(const NodeConfig& config) {
+        return updatedConfig;
+    };
+
+    auto haveProperImplementationType = [](dnnl::primitive_desc& desc, impl_desc_type implType) {
+        primitive_desc_iterator& itpd = desc;
+
+        while (itpd) {
+            const impl_desc_type descImplType = parse_impl_name(itpd.impl_info_str());
+
+            if (descImplType == implType) {
+                return true;
+            }
+
+            if (!itpd.next_impl())
+                break;
+        }
+
+        return false;
+    };
```
Move to some utility file

EgorDuplensky replied:
Done

EgorDuplensky commented on src/plugins/intel_cpu/src/nodes/conv.cpp:
```diff
@@ -912,143 +964,149 @@ void Convolution::addLegacyZeroPoints(dnnl::primitive_attr& attr) {
-                    dataConfig.setMemDesc(getSrcMemDesc(itpd, j));
-                    cfg.inConfs.push_back(dataConfig);
-                }
+    auto haveProperImplementationType = [](dnnl::primitive_desc& desc, impl_desc_type implType) {
```
Move to some utility file

EgorDuplensky replied:
Done

dmitry-gorokhov commented on src/plugins/intel_cpu/src/nodes/conv.cpp:
```diff
@@ -650,13 +670,19 @@ void Convolution::setPostOps(dnnl::primitive_attr& attr,
             if (useLegacyPostOps) {
+                bool forceBinary = false;
+                // Temporary debug functionality to be able to force binary postops for any model
+                CPU_DEBUG_CAP_ENABLE(if (std::getenv("OV_CPU_FORCE_FQ_AS_BINARY")) {forceBinary = true;})
```
Do we still need this?

EgorDuplensky replied:
I would propose to keep it for debugging for now.
This is only temporarily, I am pretty sure after migration is done till the release we need to have a flexible way to tweak those heuristics.
Can be removed before release of course.

dmitry-gorokhov replied:
Considering this is hided under Debug Caps I am ok to leave as is. However we have to either delete it once migration period is over or properly document inside corresponding developer doc page

EgorDuplensky replied:
Yes, if we want to keep it even for release, than we should implement it in proper way, like any other debug caps environment option.

dmitry-gorokhov commented on src/plugins/intel_cpu/src/nodes/kernels/registers_pool.hpp:
```diff
@@ -300,49 +298,50 @@ class IsaRegistersPool<x64::avx512_core> : public RegistersPool {
+            return std::make_shared<IsaRegistersPool<dnnl::impl::cpu::x64::avx512_core>>(regsToExclude);
+        case dnnl::impl::cpu::x64::avx512_core_amx: return std::make_shared<IsaRegistersPool<dnnl::impl::cpu::x64::avx512_core>>(regsToExclude);
+        case dnnl::impl::cpu::x64::avx512_vpopcnt: return std::make_shared<IsaRegistersPool<dnnl::impl::cpu::x64::avx512_core>>(regsToExclude);
+        case dnnl::impl::cpu::x64::isa_undef:
```
maybe just `default:` instead of explicit list below?

EgorDuplensky replied:
Compiler is not happy when we forget about enum entries in scope of switch-case

dmitry-gorokhov commented on src/plugins/intel_cpu/src/nodes/conv.cpp:
```diff
@@ -1599,27 +1677,34 @@ void Convolution::appendZeroPointsArgs() {
+        setPostOps(attr, outputStaticShape(), false);
+
+        if (attrContainsPostOp(attr, dnnl::impl::primitive_kind::binary) &&
+            attrContainsAnyOfPostOps(attr, {dnnl::impl::primitive_kind::eltwise})) {
```
Why this condition has been changed?

EgorDuplensky replied:
The original condition was relying on output scales which do not really exist anymore.
And they are basically replaced with binary post ops.
So if we keep the original condition, we never "try" brgemm conv then.
I tried many new heuristics and this one "to try brgemm conv only if there are only binary post ops" (and basically use depthwise otherwise) was the best so far. But as was discussed it still can be improved.

dmitry-gorokhov replied:
thanks for explanation. As long as we have separate activity that refines the logic - lets leave as is.

dmitry-gorokhov commented on src/plugins/intel_cpu/src/nodes/conv.cpp:
```diff
@@ -912,143 +970,149 @@ void Convolution::addLegacyZeroPoints(dnnl::primitive_attr& attr) {
+    // attr[0] for legacy post ops;
+    // attr[1] is mostly for binaryPostops except when having per-tensor zp on AMX.
+    const int descId = descIdx[selectedPrimitiveDescriptorIndex];
+    (void) descId;
```
why this line is required?

EgorDuplensky replied:
Will be updated

EgorDuplensky replied:
Done

EgorDuplensky commented on src/plugins/intel_cpu/src/dnnl_extension_utils.cpp:
```diff
@@ -150,10 +153,17 @@ std::shared_ptr<DnnlBlockedMemoryDesc> DnnlExtensionUtils::makeUndefinedDesc(con
-    return DnnlExtensionUtils::makeDescriptor(*cdesc);
+
+    dnnl_memory_desc_t cloned_md = nullptr;
+    dnnl_memory_desc_clone(&cloned_md, cdesc);
```
Since 3.0 oneDNN's C++ api structures are only holding the references (actually the pointers) to underlying C api structures.
So every time we want to copy desc we have to perform deep copy and in this case we need to clone underlying C api structure using heap (there is "new" call inside this clone) and the wrap it into memory::desc, so this memory::desc is responsible for de-allocation now.
I have also added a copy constructor for memory::desc itself which is basically doing the same (deep copy).

dmitry-gorokhov replied:
As our measurements showed - such kind of additional clone in query routine lead to performance overheads most damaging dynamic shapes performance. We need to find solution which will allow to avoid deep copy of md here. However that can be done in follow up PR.

EgorDuplensky replied:
After more investigation it turns out we can avoid this extra clone.
We just need to stick with C dnnl_memory_desc structure and pass it though until we assign it within our MemoryDesc wrapper classes.

EgorDuplensky commented on src/plugins/intel_cpu/src/graph.cpp:
```diff
@@ -67,7 +67,7 @@ typedef std::unordered_set<EdgePtr> edge_cluster_t;
 
 Graph::~Graph() {
-    CPU_DEBUG_CAP_ENABLE(summary_perf(*this));
+    // CPU_DEBUG_CAP_ENABLE(summary_perf(*this));
```
This actually breaks some tests (which are not part of the smoke_ skope)
I have already created a ticket 103441
Will revert for now

EgorDuplensky replied:
Done

EgorDuplensky commented on src/plugins/intel_cpu/src/node.h:
```diff
@@ -579,7 +579,7 @@ class Node {
     std::unordered_map<int, MemoryPtr> postOpsArgs;
     dnnl::primitive prim;
-    std::vector<DnnlDesriptor> descs;
+    std::vector<std::shared_ptr<dnnl::primitive_desc>> descs;
```
Are you sure it is? I cannot find such definition

EgorDuplensky replied:
Discussed, shared_ptr wrapper will be removed

EgorDuplensky replied:
Done

EgorDuplensky commented on src/plugins/intel_cpu/src/dnnl_postops_composer.cpp:
```diff
@@ -158,7 +164,7 @@ bool DnnlPostOpsComposer::appendScale(const std::vector<float>& scale, bool allo
     // final fallback
     if (scale.size() == 1) {
-        ops.append_eltwise(1.0f, dnnl::algorithm::eltwise_linear, scale[0], 0);
+        appendEltwise(dnnl::algorithm::eltwise_linear, scale[0], 0);
```
Will be updated

EgorDuplensky replied:
Done

EgorDuplensky commented on src/plugins/intel_cpu/src/emitters/jit_snippets_emitters.hpp:
```diff
@@ -73,6 +69,8 @@ class jit_container_emitter: public jit_emitter {
 /// }
 /// Note that Kernel doesn't accept any input arguments.
 ///
+// using namespace dnnl::impl::cpu::x64;
```
Will be updated

EgorDuplensky replied:
Done

EgorDuplensky commented on src/plugins/intel_cpu/src/memory_desc/dnnl_blocked_memory_desc.cpp:
```diff
@@ -4,7 +4,15 @@
 #include <dnnl_types.h>
+#include <oneapi/dnnl/dnnl.h>
 #include <common/memory_desc_wrapper.hpp>
+#include <oneapi/dnnl/dnnl.hpp>
```
Will be updated

EgorDuplensky replied:
Done

EgorDuplensky commented on src/plugins/intel_cpu/src/memory_desc/dnnl_memory_desc.h:
```diff
@@ -17,37 +19,35 @@ using DnnlMemoryDescCPtr = std::shared_ptr<const DnnlMemoryDesc>;
+    bool hasEmptyExtraData() const;
 
-    MemoryDescPtr cloneWithNewPrecision(const InferenceEngine::Precision prec) const override;
+    void reset(dnnl_memory_desc_t md) {
```
Will be updated

EgorDuplensky replied:
Done

EgorDuplensky commented on src/plugins/intel_cpu/src/node.cpp:
```diff
@@ -268,45 +270,62 @@ void Node::selectPreferPrimitiveDescriptor(const std::vector<impl_desc_type>& pr
+            const size_t descInConfSize = supportedPrimitiveDesc.getConfig().inConfs.size();
+
+            if (descInConfSize > getParentEdges().size()) {
+                DEBUG_LOG(getName(), " Desc ", i, " with type: ", supportedType,
```
Not sure, we can try
Will be updated

EgorDuplensky commented on src/plugins/intel_cpu/src/nodes/pooling.cpp:
```diff
@@ -78,44 +84,55 @@ struct PoolingKey {
-                                                                 const std::vector<ptrdiff_t>& effective_pad_end,
-                                                                 const std::vector<ptrdiff_t>& effective_dilation,
-                                                                 const std::vector<ptrdiff_t>& data_pad_end) {
+std::shared_ptr<dnnl::pooling_forward::primitive_desc> createDescriptorHelper(const dnnl::engine& engine,
```
Will be updated

EgorDuplensky replied:
Done

EgorDuplensky commented on src/plugins/intel_cpu/src/nodes/pooling.cpp:
```diff
@@ -431,23 +449,29 @@ dnnl::algorithm Pooling::getPoolingAlgorithm() const {
+    const dnnl::algorithm alg) {
+
+    dnnl::primitive_attr attr;
+    setPostOps(attr);
```
Will be updated

EgorDuplensky replied:
Done

EgorDuplensky commented on src/plugins/intel_cpu/src/nodes/conv.cpp:
```diff
@@ -488,7 +496,13 @@ void Convolution::getSupportedDescriptors() {
+    memory::format_tag nCsp16c = ndims == 3 ? memory::format_tag::nCw16c : (ndims == 4 ? memory::format_tag::nChw16c : memory::format_tag::nCdhw16c);
+
     if (canBeExecutedInInt8()) {
+        DEBUG_LOG("Creating I8 descriptor");
```
Will be updated

EgorDuplensky replied:
Done

EgorDuplensky commented on src/plugins/intel_cpu/src/nodes/conv.cpp:
```diff
@@ -609,15 +622,22 @@ void Convolution::setPostOps(dnnl::primitive_attr& attr,
                     break;
                 }
-                ops.append_sum(1.0, DnnlExtensionUtils::IEPrecisionToDataType(eltwisePrecision));
+                DEBUG_LOG("Append ", node->getName(), " as sum post op");
```
Will be updated

EgorDuplensky replied:
Done

EgorDuplensky commented on src/plugins/intel_cpu/src/nodes/softmax.cpp:
```diff
@@ -130,13 +131,23 @@ void SoftMax::initOptimalPrimitiveDescriptor() {
 
-    DnnlDesriptor desc(std::shared_ptr<softmax_forward::desc>(
-            new softmax_forward::desc(prop_kind::forward_scoring, in_candidate, axis)));
+    dnnl::primitive_attr attr;
```
Done

EgorDuplensky commented on src/plugins/intel_cpu/src/memory_desc/dnnl_blocked_memory_desc.cpp:
```diff
@@ -615,27 +634,32 @@ void DnnlBlockedMemoryDesc::initStrides() {
 }
 
 void DnnlBlockedMemoryDesc::recomputeDefaultStrides() {
+    IE_THROW() << "ONEDNN_3_0: Strides recomputation is not implemented" << "\n";
```
Done

EgorDuplensky commented:
At least one more modification to be applied to oneDNN branch (a fix for sparsity)

dmitry-gorokhov commented on src/plugins/intel_cpu/src/utils/ngraph_transformation.hpp:
```diff
@@ -18,7 +18,7 @@ class TransformationDumper {
     explicit TransformationDumper(const DebugCapsConfig& config, const DebugCapsConfig::TransformationFilter::Type type,
                                   const std::shared_ptr<ov::Model>& model)
-        : config(config), type(type), model(model) {
+        : config(config), model(model), type(type) {
```
Yeah, correct. But now it is not aligned with TransformationDumper ctor arguments list order. This is definetelly not important, just catches the eye. Not necessary to fix.

dmitry-gorokhov reviewed (approved):
Lets include sparsity fixes and merge

usstq commented:
@EgorDuplensky  there is a ticket for enabling fp16 (CVS-92294) which depends on [`foward, f16` ](https://github.com/oneapi-src/oneDNN/blob/f813d379eab60749a2f1f46a30a6f4eedae89046/src/cpu/cpu_convolution_list.cpp#LL157C11-L157C28), I saw this part was removed by Cc feature, should we bring them back?

dmitry-gorokhov commented:
> @EgorDuplensky there is a ticket for enabling fp16 ([CVS-92294](https://jira.devtools.intel.com/browse/CVS-92294)) which depends on [`foward, f16` ](https://github.com/oneapi-src/oneDNN/blob/f813d379eab60749a2f1f46a30a6f4eedae89046/src/cpu/cpu_convolution_list.cpp#LL157C11-L157C28), I saw this part was removed by Cc feature, should we bring them back?

I would prefer to keep this functionality removed from the binaries until FP16 is not enabled for the plugin in general. Allows to prevent unnecessary binary size increase for the upcoming release.

EgorDuplensky commented:
@dmitry-gorokhov 
* Rebased
* Sparsity fixes included, all matmul sparse tests are passed.
* Commit with fix for building oneDNN with conventional tbb is dropped (since there are no issues with building openvino and benchdnn target)
* Added constructors for **DnnlBlockedMemoryDesc** and **DnnlMemoryDesc** which accept pointer to C dnnl_memory_desc data structure, which helps to avoid and extra copy for **query_md**

antonvor reviewed (approved):
sparse decompression feature related changes approved

EgorDuplensky commented:
@usstq 
Could you please take a look at the last commit of the PR

usstq reviewed (approved):
no accuracy concerns from my side, performance drop is inevitable since `can_fuse_into_oscale` is not working anymore. but may be we can fix it in the future.

ilya-lavrenov commented on src/inference/include/openvino/runtime/intel_cpu/properties.hpp:
```diff
@@ -47,7 +47,21 @@ namespace intel_cpu {
+ * plugin at the model compilation stage and store non-zero values in a special packed format. Then, during the
+ * execution of the model, the weights are unpacked and used in the computational kernel. Since the weights are loaded
+ * from DDR/L3 cache in the packed format this significantly decreases memory consumption and as a consequence improve
+ * inference performance. The following code allows to set the sparse rate value.
```
secondly, I don't see that this config operation is reported as supported_property by plugin.

```
[ INFO ] Available devices: 
[ INFO ] CPU
[ INFO ] 	SUPPORTED_PROPERTIES: 
[ INFO ] 		Immutable: AVAILABLE_DEVICES : ""
[ INFO ] 		Immutable: RANGE_FOR_ASYNC_INFER_REQUESTS : 1 1 1
[ INFO ] 		Immutable: RANGE_FOR_STREAMS : 1 1
[ INFO ] 		Immutable: FULL_DEVICE_NAME : VirtualApple @ 2.50GHz
[ INFO ] 		Immutable: OPTIMIZATION_CAPABILITIES : FP32 FP16 INT8 BIN EXPORT_IMPORT
[ INFO ] 		Immutable: CACHING_PROPERTIES : ""
[ INFO ] 		Mutable: NUM_STREAMS : 1
[ INFO ] 		Mutable: AFFINITY : CORE
[ INFO ] 		Mutable: INFERENCE_NUM_THREADS : 0
[ INFO ] 		Mutable: PERF_COUNT : NO
[ INFO ] 		Mutable: INFERENCE_PRECISION_HINT : f32
[ INFO ] 		Mutable: PERFORMANCE_HINT : UNDEFINED
[ INFO ] 		Mutable: PERFORMANCE_HINT_NUM_REQUESTS : 0
[ INFO ] 

```

antonvor replied:
@ilya-lavrenov we store weights in compressed format inside CPU's implementation
I can add this property to report in separate PR

antonvor replied:
PR https://github.com/openvinotoolkit/openvino/pull/16569
